mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java

```diff
@@ -183,6 +187,82 @@ public void setUp() throws Exception {
+
+        assertThat(messageSearchIndex.search(session,
+            MultimailboxesSearchQuery.from(searchQuery)
+                .inMailboxes(mailbox.getMailboxId(), mailbox2.getMailboxId())
+                .build(), 20))
+            .hasSize(12)
+            .containsOnly(m1.getMessageId(),
+                m2.getMessageId(),
+                m3.getMessageId(),
+                m4.getMessageId(),
+                m5.getMessageId(),
+                m6.getMessageId(),
+                m7.getMessageId(),
+                m8.getMessageId(),
+                m9.getMessageId(),
+                mOther.getMessageId(),
+                mailWithAttachment.getMessageId(),
+                mailWithInlinedAttachment.getMessageId());
+    }
+
+    @Test
+    public void searchingMessageInMultipleMailboxShouldUnionOfTheTwoMailbox() throws MailboxException {
+        Assume.assumeTrue(messageIdManager != null);
+        messageIdManager.setInMailboxes(m4.getMessageId(),
+            ImmutableList.of(mailbox2.getMailboxId()),
+            session);
+
+        SearchQuery searchQuery = new SearchQuery();
+
+        assertThat(messageSearchIndex.search(session,
+            MultimailboxesSearchQuery.from(searchQuery)
+                .inMailboxes(mailbox.getMailboxId(), mailbox2.getMailboxId())
+                .build(), 20))
+            .containsOnly(m1.getMessageId(),
+                m2.getMessageId(),
+                m3.getMessageId(),
+                m4.getMessageId(),
+                m5.getMessageId(),
+                m6.getMessageId(),
+                m7.getMessageId(),
+                m8.getMessageId(),
+                m9.getMessageId(),
+                mOther.getMessageId(),
+                mailWithAttachment.getMessageId(),
+                mailWithInlinedAttachment.getMessageId());
+    }
+
+    @Test
+    public void searchingMessageInMultipleMailboxShouldNotReturnLessMessageThanLimitArgument() throws MailboxException {
+        Assume.assumeTrue(messageIdManager != null);
+        messageIdManager.setInMailboxes(m1.getMessageId(), ImmutableList.of(mailbox.getMailboxId(), mailbox2.getMailboxId()), session);
+        messageIdManager.setInMailboxes(m2.getMessageId(), ImmutableList.of(mailbox.getMailboxId(), mailbox2.getMailboxId()), session);
+        messageIdManager.setInMailboxes(m3.getMessageId(), ImmutableList.of(mailbox.getMailboxId(), mailbox2.getMailboxId()), session);
+        messageIdManager.setInMailboxes(m4.getMessageId(), ImmutableList.of(mailbox.getMailboxId(), mailbox2.getMailboxId()), session);
+        messageIdManager.setInMailboxes(m5.getMessageId(), ImmutableList.of(mailbox.getMailboxId(), mailbox2.getMailboxId()), session);
+        messageIdManager.setInMailboxes(m6.getMessageId(), ImmutableList.of(mailbox.getMailboxId(), mailbox2.getMailboxId()), session);
+        messageIdManager.setInMailboxes(m7.getMessageId(), ImmutableList.of(mailbox.getMailboxId(), mailbox2.getMailboxId()), session);
```

Comment:
This test covers the fact that message in both mailboxes will get the limmit wrong (deduplication after limit)

If your test is red but this one green, an other issue is going on !

Hmmm Maybe we can introduce it again....